synapse/handlers/federation_event.py

```diff
@@ -792,9 +792,41 @@ async def _process_pulled_events(
             ],
         )
 
+        # Check if we already any of these have these events.
+        # Note: we currently make a lookup in the database directly here rather than
+        # checking the event cache, due to:
+        # https://github.com/matrix-org/synapse/issues/13476
+        existing_events_map = await self._store._get_events_from_db(
+            [event.event_id for event in events]
+        )
+
+        new_events = []
+        for event in events:
+            event_id = event.event_id
+
+            # If we've already seen this event ID...
+            if event_id in existing_events_map:
+                existing_event = existing_events_map[event_id]
+
+                # ...and the event itself was not previously stored as an outlier...
+                if not existing_event.event.internal_metadata.is_outlier():
+                    # ...then there's no need to persist it. We have it already.
+                    logger.info(
+                        "_process_pulled_event: Ignoring received event %s which we "
+                        "have already seen",
+                        event.event_id,
+                    )
+
+                # While we have seen this event before, it was stored as an outlier.
+                # We'll now persist it as a non-outlier.
+                logger.info("De-outliering event %s", event_id)
```

Comment:
This control flow is only doing logging! I think you wanted a `continue` in the inner if block, right? :-)

Reply:
(Maybe this indicates that a test is needed...)

Reply:
Serves me right for not manually testing again before putting this up!

I'll write up a unit test for this function, but otherwise I find this behaviour a [little hard to test](https://github.com/matrix-org/synapse/pull/14161#issuecomment-1276527040).

Reply:
A complement test seems like it would fit https://github.com/matrix-org/synapse/pull/14161#issuecomment-1276527040 well

If you're looking for a test that works with `_process_pulled_events`, I wrote one in https://github.com/matrix-org/synapse/pull/13864, see https://github.com/matrix-org/synapse/blob/3f8fef2beb2195d2220b3af873daee94971d3e8a/tests/handlers/test_federation_event.py#L872-L1320

Reply:
The problem with a Complement test is that we'd need to call Synapse's [Delete Room Admin API](https://matrix-org.github.io/synapse/latest/admin_api/rooms.html#delete-room-api), which would only be specific to Synapse. I could write this test, but I wouldn't be able to publish it anywhere permanent, and the steps in it would be no different than manually testing.

Manually testing again, with the `continue` backfill doesn't drop any events. Thanks for catching that :)

Reply:
I cracked and settled for writing a unit test for backfill, while including a bit that specifically checks that backfill won't rely on the event cache to determine whether to persist an event. The result is a test for the backfill + the specific behaviour that PR removes.

https://github.com/matrix-org/synapse/blob/42f14104e36b54d7e64d7543c6835444e340c5cc/tests/handlers/test_federation.py#L332-L426